### DISCUSSION TITLE
Unable To resolve Maximum Capacitance Violations + Understanding metada-base-ok.json

### DISCUSSION THREAD

**Discussion by dalyles** [The-OpenROAD-Project/OpenROAD-flow-scripts#1120] [General] [answered]
Hello,

I'm running OpenROAD on the IBEX example file and I have come across maximum capacitance violations.

Below is a screenshot of my <code>constraints.sdc</code> file.
The default value for the clock period was 2.8ns but I set it to 3ns to overcome a setup time violation
that occured in a previous run. I also changed the clock period in <code>metadata-base-ok.json</code> to 3ns
in order to match with the clock period value in <code>constraints.sdc</code>.
<img width="605" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/efdb9b63-48f3-4bba-a420-3732c863b36a">

<img width="541" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/af554115-6896-4a8f-99db-2790cb26cc77">

The main issue that I have here is simply, in what ways can I remedy the maximum capacitance violations?
<img width="390" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/d41b8aa0-4638-41dd-b419-e9b554488278">

One approach I did was to go to to the <code>metadata-base-ok.json</code> file and try to increase the <code>"finish__timing__drv__max_cap_limit"</code>  from -0.0575862 to 0.20. Unfortunately the result is
still the same.

I also have some further questions:
1) What parameter(s) in the <code>metadata-base-ok.json</code> should be tuned to fix the max cap violations?
2) Were the parameter values in <code>metadata-base-ok.json</code> set up by a human or automatically generated?
3) What was the reasoning behind choosing such values in <code>metadata-base-ok.json</code>?
4) Is there any supporting documentation that explains what the parameters mean in <code>metadata-base-ok.json</code>?

Thank you so much for anyone who's able to help!

**Comment by vijayank88** [comment]
@dalyles 
Nothing to do with `metadata-base-ok.json`. Its just an metric report file.
`config.mk` & `constraint.sdc` is the files you've to tune to get better PPA.
Did you referred this flow [tutorial](https://openroad-flow-scripts.readthedocs.io/en/latest/tutorials/FlowTutorial.html#introduction)?
About metrics you find [here](https://openroad-flow-scripts.readthedocs.io/en/latest/contrib/Metrics.html).
Flow tutorial videos available [here](https://theopenroadproject.org/videos/).

**Reply by dalyles** [reply]
Hi @vijayank88 

Thanks for your reply. 
I looked at the resources that you provided me and I wanted to check with you
if installing the autotuner would be the right direction to go in order to resolve
maximum cap violations? 

From what I understand, using the autotuner would help generate a design with no
violations in general, right?

Looking forward to your reply.

**Comment by maliberty** [comment]
You could try adding -slew_margin <value> to repair_design to force it to overfix and see if that helps.  The larger the value the more the overfixing.

**Reply by dalyles** [reply]
Hi @maliberty 

Thank you for adding your input. 
Just for context, is this command used when running the GUI?
I'm currently running OpenROAD using <code>make</code>.
So withing the context of running <code>make</code> which file
should I be modifying to add <code>-slew_margin</code> to <code>repair_design</code>?

Thanks!

**Reply by maliberty** [reply]
This command would need to be run as part of your flow.  I don't know if you are using OL or ORFS or another.

**Reply by dalyles** [reply]
Oh yes, I'm using ORFS for my runs if that helps.

**Comment by maliberty** [comment]
We don't have a control variable but you would add it to https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/ec6bc0a349fa1f315c9debc2b1ad8633fc45030a/flow/scripts/resize.tcl#L36

If that helps we can add a configuration variable.

**Reply by dalyles** [reply]
I think if you all add it that would be great, so that others won't have to come across this issue in the future.
Also what do you mean exactly by configuration variable? How does that differ from the control variable that you mentioned?

**Comment by maliberty** [comment]
I just noticed that you have max cap violations not max slew so you would use -cap_margin <value>

**Reply by dalyles** [reply]
No worries, your previous answer is also very useful because in other runs I also encounter slew rate violations.

**Reply by dalyles** [reply]
Runs for other circuits that is.

**Comment by maliberty** [comment]
Something like https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/ec6bc0a349fa1f315c9debc2b1ad8633fc45030a/flow/scripts/cts.tcl#L66

where you set SETUP_SLACK_MARGIN in your config.mk and the flow picks it up without having to  modify the flow scripts.

**Comment by dalyles** [comment]
Just for clairty, in the line code given below,

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/ec6bc0a349fa1f315c9debc2b1ad8633fc45030a/flow/scripts/resize.tcl#L36

The new line would replaced with <code>repair_design -cap_margin -slew_margin</code>?

**Reply by maliberty** [reply]
Some like repair_design -cap_margin 10 -slew_margin 10

**Reply by dalyles** [reply]
Just to follow up, is 10 considered a high margin?

**Reply by maliberty** [reply]
They are percentage arguments so that should be 10% over-fixing.   I think you'll have to experiment a bit.  We do have some work underway that should help this issue in general.

**Comment by dalyles** [comment]
Great!
Thanks a lot @maliberty, just by setting the margins to 10 percent for -cap_margin,
I was able to reduce the cap violations down from 18 to 5. I'll continue to play with the values until I can get
it the violation count to converge to zero.
<img width="550" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/ed5e6df3-4ea4-42de-ba86-280186c879ea">

**Comment by maliberty** [comment]
I'm adding the vars in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1122

**Comment by dalyles** [comment]
Thanks @maliberty 

Is this snipped of code still required though?
<code>repair_design -cap_margin 10 -slew_margin 10</code>

I tried running the updated version of your file without this line of code and
it came back to the original issue of having 18 max cap violations.

**Reply by maliberty** [reply, marked as answer]
You can set in your config.mk
```
export CAP_MARGIN = 10
export SLEW_MARGIN = 10
```
and not have to mess with resize.tcl

**Reply by vijayank88** [reply]
The default margin is 0. How do we decide how much margin to be used to fix the violations?
Is there any guidelines or rules to calculate `MARGIN` to overfix it? @maliberty

**Reply by maliberty** [reply]
Its mostly experimental.  We are working on post-grt fixing which should be more accurate as it is near the end of the of flow.
